canonical/ght
# Add recording of Puppeteer session

## Rationale
When something goes wrong with any given GHT operation, it is often quite difficult to troubleshoot from our end because we cannot see exactly where Puppeteer failed. This PR is to add an optional `--record` flag that will record the Puppeteer session and output to `~/.ght/recordings`. Users can then supply this video along with issue reports to WPE to help us better find and fix issues.

The screen recording is done using an NPM package called [puppeteer-screen-recorder](https://prasanaworld.github.io/puppeteer-screen-recorder/index.html).

## Done
- Added a `ScreenRecorder` class which handles creating the `PuppeteerScreenRecorder` instance, passing it the Puppeteer `page` object, defining output save paths and filenames, and starting/stopping a recording.
  - **N.B.** The recording is saved to an output stream. This is to ensure that we still save the recording in the event of an error. This functionality is most likely to be used in the event of an error occurring. 
- Added an optional `--record` flag which will trigger recording a session.
- Updated the `getPuppeteer()` function in the `BaseController` to optionally start a recording whenever a Puppeteer session is created. This only happens if the `--record` flag is set.
- Added a `stopRecording` function to the `BaseController` to allow any given controller to stop its existing recording.
- Wrapped the `run()` function of each controller in a try catch, to allow stopping the recording in the finally block. This way, recordings will always be stopped even in the event of an error.
- Updated the README to give users instructions on how to record a session.
- Updated snapcraft.yaml to install ffmpeg + required packages, and add package locations to `LD_LIBRARY_PATH` var

## QA
- Checkout this feature branch
- ### Test Locally
  - Run GHT locally, adding the `--record` flag to see if a recording is saved as expected. Since we save to an output stream, you don't need to complete any given GHT operation. You should be able to exit early and still see a video saved.
  - Here is an example command with the assign functionality:
  ```
  yarn dev assign -i --record
  ```
  - Ensure that a video of your session is saved, and that you can view the video with no issues.
- ### Test as Snap
  - Temporarily update `name` field in snapcraft.yaml to `ght-dev`
  - Build snap: 
  ```
  snapcraft
  ```
  -  Install snap:
  ```
  sudo snap install --dangerous --devmode ght-dev_1.11.0_amd64.snap
  ```
  -  Run a GHT command with record flag:
  ```
  ght-dev.ght assign -i --new --record
  ```

## Issue / Card
[WD-14846](https://warthogs.atlassian.net/browse/WD-14846)


[WD-14846]: https://warthogs.atlassian.net/browse/WD-14846?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
